File: rust/src/lib/rate_limiter.rs
// Rate Limiter for Moltbook API
//
// TWO LEVELS OF RATE LIMITING:
// 1. GLOBAL (IP-level) - Shared across ALL agents
// 2. PER-AGENT - Each agent has their own limits
//
// Rust port of TypeScript rateLimiter

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};
use once_cell::sync::Lazy;

use crate::constants::*;
use crate::types::RateLimitState;

// =============================================================================
// GLOBAL RATE LIMITING (IP-level)
// =============================================================================

const GLOBAL_REQUEST_WINDOW_MS: i64 = 60 * 1000;  // 1 minute
const GLOBAL_POST_WINDOW_MS: i64 = 60 * 60 * 1000;  // 1 hour

#[derive(Debug, Clone)]
struct GlobalRateLimitState {
    requests: Vec<i64>,
    posts: Vec<i64>,
    retry_after: Option<i64>,
}

impl Default for GlobalRateLimitState {
    fn default() -> Self {
        Self {
            requests: Vec::new(),
            posts: Vec::new(),
            retry_after: None,
        }
    }
}

static GLOBAL_STATE: Lazy<Arc<Mutex<GlobalRateLimitState>>> = 
    Lazy::new(|| Arc::new(Mutex::new(GlobalRateLimitState::default())));

fn current_time_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_millis() as i64
}

fn can_make_global_request() -> bool {
    let mut state = GLOBAL_STATE.lock().unwrap();
    
    // Check retry-after
    if let Some(retry_after) = state.retry_after {
        if current_time_ms() < retry_after {
            return false;
        }
    }
    
    // Prune old requests
    let cutoff = current_time_ms() - GLOBAL_REQUEST_WINDOW_MS;
    state.requests.retain(|&t| t > cutoff);
    
    state.requests.len() < GLOBAL_REQUESTS_PER_MIN
}

fn can_post_globally() -> bool {
    let mut state = GLOBAL_STATE.lock().unwrap();
    
    // Prune old posts
    let cutoff = current_time_ms() - GLOBAL_POST_WINDOW_MS;
    state.posts.retain(|&t| t > cutoff);
    
    state.posts.len() < GLOBAL_POSTS_PER_HOUR
}

fn record_global_request() {
    let mut state = GLOBAL_STATE.lock().unwrap();
    state.requests.push(current_time_ms());
}

fn record_global_post() {
    let mut state = GLOBAL_STATE.lock().unwrap();
    state.posts.push(current_time_ms());
}

pub fn set_global_retry_after(retry_after_seconds: i64) {
    let mut state = GLOBAL_STATE.lock().unwrap();
    state.retry_after = Some(current_time_ms() + retry_after_seconds * 1000);
}

// =============================================================================
// PER-AGENT RATE LIMITING
// =============================================================================

type AgentStates = Arc<Mutex<HashMap<String, RateLimitState>>>;

static AGENT_STATES: Lazy<AgentStates> = 
    Lazy::new(|| Arc::new(Mutex::new(HashMap::new())));

fn get_agent_state(agent_id: &str) -> RateLimitState {
    let mut states = AGENT_STATES.lock().unwrap();
    states
        .entry(agent_id.to_string())
        .or_insert_with(RateLimitState::default)
        .clone()
}

fn update_agent_state(agent_id: &str, state: RateLimitState) {
    let mut states = AGENT_STATES.lock().unwrap();
    states.insert(agent_id.to_string(), state);
}

fn prune_old_entries(entries: &[crate::types::RateLimitRequest], window_ms: i64) -> Vec<crate::types::RateLimitRequest> {
    let cutoff = current_time_ms() - window_ms;
    entries
        .iter()
        .filter(|e| e.timestamp > cutoff)
        .cloned()
        .collect()
}

pub fn can_make_request(agent_id: &str) -> bool {
    // Check global first
    if !can_make_global_request() {
        return false;
    }
    
    // Check per-agent
    let mut state = get_agent_state(agent_id);
    
    // Check retry-after
    if let Some(retry_after) = state.retry_after {
        if current_time_ms() < retry_after {
            return false;
        }
    }
    
    // Prune old requests
    state.requests = prune_old_entries(&state.requests, RATE_LIMIT_REQUEST_WINDOW_MS);
    update_agent_state(agent_id, state.clone());
    
    state.requests.len() < RATE_LIMIT_REQUESTS_PER_MIN
}

pub fn record_request(agent_id: &str) {
    record_global_request();
    
    let mut state = get_agent_state(agent_id);
    state.requests.push(crate::types::RateLimitRequest {
        timestamp: current_time_ms(),
    });
    update_agent_state(agent_id, state);
}

pub fn can_post(agent_id: &str) -> bool {
    if !can_make_request(agent_id) {
        return false;
    }
    
    if !can_post_globally() {
        return false;
    }
    
    let state = get_agent_state(agent_id);
    
    // Check last post time
    if let Some(last_post) = state.posts.last() {
        let elapsed = current_time_ms() - last_post.timestamp;
        if elapsed < RATE_LIMIT_POST_INTERVAL_SEC * 1000 {
            return false;
        }
    }
    
    true
}

pub fn record_post(agent_id: &str) {
    record_global_post();
    record_request(agent_id);
    
    let mut state = get_agent_state(agent_id);
    state.posts.push(crate::types::RateLimitRequest {
        timestamp: current_time_ms(),
    });
    update_agent_state(agent_id, state);
}

pub fn can_comment(agent_id: &str) -> bool {
    if !can_make_request(agent_id) {
        return false;
    }
    
    let mut state = get_agent_state(agent_id);
    state.comments = prune_old_entries(&state.comments, RATE_LIMIT_COMMENT_WINDOW_MS);
    update_agent_state(agent_id, state.clone());
    
    state.comments.len() < RATE_LIMIT_COMMENTS_PER_HOUR
}

pub fn record_comment(agent_id: &str) {
    record_request(agent_id);
    
    let mut state = get_agent_state(agent_id);
    state.comments.push(crate::types::RateLimitRequest {
        timestamp: current_time_ms(),
    });
    update_agent_state(agent_id, state);
}

pub fn get_time_until_can_post(agent_id: &str) -> i64 {
    let state = get_agent_state(agent_id);
    
    if let Some(last_post) = state.posts.last() {
        let elapsed = current_time_ms() - last_post.timestamp;
        let required = RATE_LIMIT_POST_INTERVAL_SEC * 1000;
        return std::cmp::max(0, required - elapsed);
    }
    
    0
}

File: typescript/src/lib/rateLimiter.ts
/**
 * Rate Limiter for Moltbook API
 *
 * TWO LEVELS OF RATE LIMITING:
 *
 * 1. GLOBAL (IP-level) - Shared across ALL agents in this process
 *    WHY? Many agents from one IP could trigger Moltbook's IP-level blocks.
 *    We limit total requests to a safe global maximum.
 *
 * 2. PER-AGENT - Each agent has their own Moltbook account limits
 *    WHY? Moltbook tracks per-account rate limits (100 req/min, etc.)
 *    Each agent needs independent tracking.
 *
 * The global limiter is more conservative to ensure we don't get IP-banned
 * when running many agents from the same machine.
 */

import type { UUID } from "@elizaos/core";
import {
  GLOBAL_POSTS_PER_HOUR,
  GLOBAL_REQUESTS_PER_MIN,
  RATE_LIMIT_COMMENT_WINDOW_MS,
  RATE_LIMIT_COMMENTS_PER_HOUR,
  RATE_LIMIT_POST_INTERVAL_SEC,
  RATE_LIMIT_REQUEST_WINDOW_MS,
  RATE_LIMIT_REQUESTS_PER_MIN,
} from "../constants";
import type { AgentMoltbookState } from "../types";

// =============================================================================
// GLOBAL RATE LIMITING (IP-level protection)
// =============================================================================

/** Window for global request rate limiting (1 minute) */
const GLOBAL_REQUEST_WINDOW_MS = 60 * 1000;

/** Window for global post rate limiting (1 hour) */
const GLOBAL_POST_WINDOW_MS = 60 * 60 * 1000;

/** Global state shared across ALL agents */
interface GlobalRateLimitState {
  requests: { timestamp: number }[];
  posts: { timestamp: number }[];
  retryAfter?: number; // If Moltbook returns 429, ALL agents pause
}

const globalState: GlobalRateLimitState = {
  requests: [],
  posts: [],
};

/**
 * Check if we can make a request globally (IP-level)
 */
function canMakeGlobalRequest(): boolean {
  // Check global retry-after (if any agent got 429, all pause)
  if (globalState.retryAfter && Date.now() < globalState.retryAfter) {
    return false;
  }

  // Prune old requests
  const cutoff = Date.now() - GLOBAL_REQUEST_WINDOW_MS;
  globalState.requests = globalState.requests.filter((r) => r.timestamp > cutoff);

  return globalState.requests.length < GLOBAL_REQUESTS_PER_MIN;
}

/**
 * Check if we can post globally (IP-level)
 */
function canPostGlobally(): boolean {
  // Prune old posts
  const cutoff = Date.now() - GLOBAL_POST_WINDOW_MS;
  globalState.posts = globalState.posts.filter((p) => p.timestamp > cutoff);

  return globalState.posts.length < GLOBAL_POSTS_PER_HOUR;
}

/**
 * Record a global request
 */
function recordGlobalRequest(): void {
  globalState.requests.push({ timestamp: Date.now() });
}

/**
 * Record a global post
 */
function recordGlobalPost(): void {
  globalState.posts.push({ timestamp: Date.now() });
}

/**
 * Set global retry-after (affects ALL agents)
 */
export function setGlobalRetryAfter(retryAfterSeconds: number): void {
  globalState.retryAfter = Date.now() + retryAfterSeconds * 1000;
}

/**
 * Get global rate limit status
 */
export function getGlobalRateLimitStatus(): {
  canRequest: boolean;
  canPost: boolean;
  requestsUsed: number;
  requestsMax: number;
  postsUsed: number;
  postsMax: number;
  retryAfter?: number;
} {
  // Prune old entries
  const requestCutoff = Date.now() - GLOBAL_REQUEST_WINDOW_MS;
  const postCutoff = Date.now() - GLOBAL_POST_WINDOW_MS;
  globalState.requests = globalState.requests.filter((r) => r.timestamp > requestCutoff);
  globalState.posts = globalState.posts.filter((p) => p.timestamp > postCutoff);

  return {
    canRequest: canMakeGlobalRequest(),
    canPost: canPostGlobally(),
    requestsUsed: globalState.requests.length,
    requestsMax: GLOBAL_REQUESTS_PER_MIN,
    postsUsed: globalState.posts.length,
    postsMax: GLOBAL_POSTS_PER_HOUR,
    retryAfter: globalState.retryAfter,
  };
}

// =============================================================================
// PER-AGENT RATE LIMITING
// =============================================================================

/**
 * Per-agent state storage
 * Key: agentId, Value: agent's Moltbook state
 */
const agentStates = new Map<UUID, AgentMoltbookState>();

/**
 * Get or create state for an agent
 */
export function getAgentState(agentId: UUID): AgentMoltbookState {
  let state = agentStates.get(agentId);
  if (!state) {
    state = {
      rateLimits: {
        requests: [],
        posts: [],
        comments: [],
      },
    };
    agentStates.set(agentId, state);
  }
  return state;
}

/**
 * Clean up old entries from rate limit arrays
 */
function pruneOldEntries(
  entries: { timestamp: number }[],
  windowMs: number
): { timestamp: number }[] {
  const cutoff = Date.now() - windowMs;
  return entries.filter((e) => e.timestamp > cutoff);
}

/**
 * Check if we can make a general API request
 *
 * Checks BOTH global (IP-level) and per-agent limits.
 * Both must allow the request for it to proceed.
 */
export function canMakeRequest(agentId: UUID): boolean {
  // FIRST: Check global (IP-level) limits
  if (!canMakeGlobalRequest()) {
    return false;
  }

  // THEN: Check per-agent limits
  const state = getAgentState(agentId);

  // Check agent-specific retry-after header
  if (state.rateLimits.retryAfter && Date.now() < state.rateLimits.retryAfter) {
    return false;
  }

  // Prune old requests
  state.rateLimits.requests = pruneOldEntries(
    state.rateLimits.requests,
    RATE_LIMIT_REQUEST_WINDOW_MS
  );

  return state.rateLimits.requests.length < RATE_LIMIT_REQUESTS_PER_MIN;
}

/**
 * Record a request was made
 *
 * Records in BOTH global and per-agent tracking.
 */
export function recordRequest(agentId: UUID): void {
  // Record globally
  recordGlobalRequest();

  // Record per-agent
  const state = getAgentState(agentId);
  state.rateLimits.requests.push({ timestamp: Date.now() });
}

/**
 * Set retry-after from server response
 *
 * Sets for the specific agent, AND globally if it's a 429.
 * WHY globally? A 429 often indicates IP-level throttling,
 * so ALL agents should back off.
 */
export function setRetryAfter(agentId: UUID, retryAfterSeconds: number): void {
  // Set for this agent
  const state = getAgentState(agentId);
  state.rateLimits.retryAfter = Date.now() + retryAfterSeconds * 1000;

  // ALSO set globally - if one agent gets 429, all should pause
  // This protects against IP-level bans
  setGlobalRetryAfter(retryAfterSeconds);
}

/**
 * Check if we can make a post
 *
 * Checks BOTH global and per-agent post limits.
 */
export function canPost(agentId: UUID): boolean {
  // FIRST: Check if requests are allowed at all
  if (!canMakeRequest(agentId)) return false;

  // SECOND: Check global post limit (IP-level)
  if (!canPostGlobally()) return false;

  // THIRD: Check per-agent post limit
  const state = getAgentState(agentId);

  // Get the most recent post
  const lastPost = state.rateLimits.posts[state.rateLimits.posts.length - 1];
  if (lastPost) {
    const timeSinceLastPost = Date.now() - lastPost.timestamp;
    if (timeSinceLastPost < RATE_LIMIT_POST_INTERVAL_SEC * 1000) {
      return false;
    }
  }

  return true;
}

/**
 * Record a post was made
 *
 * Records in BOTH global and per-agent tracking.
 */
export function recordPost(agentId: UUID): void {
  // Record globally
  recordGlobalPost();

  // Record per-agent
  const state = getAgentState(agentId);
  state.rateLimits.posts.push({ timestamp: Date.now() });
  recordRequest(agentId);
}

/**
 * Get time until next post is allowed (in ms)
 */
export function getTimeUntilCanPost(agentId: UUID): number {
  const state = getAgentState(agentId);
  const lastPost = state.rateLimits.posts[state.rateLimits.posts.length - 1];

  if (!lastPost) return 0;

  const elapsed = Date.now() - lastPost.timestamp;
  const required = RATE_LIMIT_POST_INTERVAL_SEC * 1000;

  return Math.max(0, required - elapsed);
}

/**
 * Check if we can make a comment
 */
export function canComment(agentId: UUID): boolean {
  if (!canMakeRequest(agentId)) return false;

  const state = getAgentState(agentId);

  // Prune old comments
  state.rateLimits.comments = pruneOldEntries(
    state.rateLimits.comments,
    RATE_LIMIT_COMMENT_WINDOW_MS
  );

  return state.rateLimits.comments.length < RATE_LIMIT_COMMENTS_PER_HOUR;
}

/**
 * Record a comment was made
 */
export function recordComment(agentId: UUID): void {
  const state = getAgentState(agentId);
  state.rateLimits.comments.push({ timestamp: Date.now() });
  recordRequest(agentId);
}

/**
 * Get remaining comment quota
 */
export function getRemainingComments(agentId: UUID): number {
  const state = getAgentState(agentId);

  // Prune old comments
  state.rateLimits.comments = pruneOldEntries(
    state.rateLimits.comments,
    RATE_LIMIT_COMMENT_WINDOW_MS
  );

  return RATE_LIMIT_COMMENTS_PER_HOUR - state.rateLimits.comments.length;
}

/**
 * Get rate limit status summary for an agent
 */
export function getRateLimitStatus(agentId: UUID): {
  canRequest: boolean;
  canPost: boolean;
  canComment: boolean;
  requestsRemaining: number;
  commentsRemaining: number;
  timeUntilCanPost: number;
  retryAfter?: number;
} {
  const state = getAgentState(agentId);

  // Prune old entries
  state.rateLimits.requests = pruneOldEntries(
    state.rateLimits.requests,
    RATE_LIMIT_REQUEST_WINDOW_MS
  );
  state.rateLimits.comments = pruneOldEntries(
    state.rateLimits.comments,
    RATE_LIMIT_COMMENT_WINDOW_MS
  );

  return {
    canRequest: canMakeRequest(agentId),
    canPost: canPost(agentId),
    canComment: canComment(agentId),
    requestsRemaining: RATE_LIMIT_REQUESTS_PER_MIN - state.rateLimits.requests.length,
    commentsRemaining: RATE_LIMIT_COMMENTS_PER_HOUR - state.rateLimits.comments.length,
    timeUntilCanPost: getTimeUntilCanPost(agentId),
    retryAfter: state.rateLimits.retryAfter,
  };
}

/**
 * Clear all state for an agent (useful for testing)
 */
export function clearAgentState(agentId: UUID): void {
  agentStates.delete(agentId);
}

/**
 * Clear all agent states (useful for testing)
 */
export function clearAllStates(): void {
  agentStates.clear();
}
